File: gigvora-frontend-reactjs/src/authorization/permissionMatrix.js
import permissionMatrix from '@shared-contracts/domain/auth/permissionMatrix.json';

export function normaliseMembershipKey(value) {
  if (value == null) {
    return null;
  }
  const stringified = `${value}`.trim().toLowerCase();
  if (!stringified) {
    return null;
  }
  return stringified.replace(/[^a-z0-9]+/g, '_');
}

export function normalisePermissionKey(value) {
  if (value == null) {
    return null;
  }
  const stringified = `${value}`.trim().toLowerCase();
  return stringified || null;
}

const permissionIndex = new Map();
const permissionImplications = new Map();
const canonicalMembershipIndex = new Map();
const membershipAliasIndex = new Map();
const membershipOrder = [];

function registerPermissionImplication(source, implied) {
  const bucket = permissionImplications.get(source) ?? new Set();
  implied.forEach((permission) => {
    if (permission && permission !== source) {
      bucket.add(permission);
    }
  });
  permissionImplications.set(source, bucket);
}

(permissionMatrix.permissions ?? []).forEach((permission) => {
  const key = normalisePermissionKey(permission.key);
  if (!key || permissionIndex.has(key)) {
    return;
  }
  const implies = Array.isArray(permission.implies)
    ? permission.implies.map(normalisePermissionKey).filter(Boolean)
    : [];
  permissionIndex.set(key, {
    key,
    label: permission.label ?? key,
    description: permission.description ?? '',
    category: permission.category ?? 'general',
    surfaces: Array.isArray(permission.surfaces) ? permission.surfaces : [],
    escalationPath: Array.isArray(permission.escalationPath)
      ? permission.escalationPath.map(normaliseMembershipKey).filter(Boolean)
      : [],
    implies,
  });
  registerPermissionImplication(key, implies);
});

(permissionMatrix.memberships ?? []).forEach((membership) => {
  const canonicalKey = normaliseMembershipKey(membership.key);
  if (!canonicalKey) {
    return;
  }
  if (!canonicalMembershipIndex.has(canonicalKey)) {
    membershipOrder.push(canonicalKey);
  }
  const permissions = new Set(
    (membership.permissions ?? []).map(normalisePermissionKey).filter(Boolean),
  );
  const definition = {
    key: canonicalKey,
    label: membership.label ?? canonicalKey,
    description: membership.description ?? '',
    tier: membership.tier ?? 'standard',
    escalationPath: Array.isArray(membership.escalationPath)
      ? membership.escalationPath.map(normaliseMembershipKey).filter(Boolean)
      : [],
    grantAll: membership.grantAll === true,
    permissions,
  };
  canonicalMembershipIndex.set(canonicalKey, definition);
  const aliases = Array.isArray(membership.aliases) ? membership.aliases : [];
  const resolvedAliases = new Set([canonicalKey]);
  aliases.map(normaliseMembershipKey).filter(Boolean).forEach((alias) => resolvedAliases.add(alias));
  resolvedAliases.forEach((alias) => {
    membershipAliasIndex.set(alias, definition);
  });
});

function expandPermissionSet(permissionSet, sourceMap) {
  const queue = Array.from(permissionSet);
  while (queue.length) {
    const current = queue.pop();
    const implied = permissionImplications.get(current);
    if (!implied) {
      continue;
    }
    implied.forEach((permission) => {
      if (!permissionSet.has(permission)) {
        permissionSet.add(permission);
        queue.push(permission);
      }
      const currentSources = sourceMap.get(current);
      if (!currentSources || !currentSources.size) {
        return;
      }
      const target = sourceMap.get(permission) ?? new Set();
      currentSources.forEach((source) => target.add(source));
      sourceMap.set(permission, target);
    });
  }
}

function cloneMembership(definition) {
  if (!definition) {
    return null;
  }
  return {
    key: definition.key,
    label: definition.label,
    description: definition.description,
    tier: definition.tier,
    escalationPath: [...definition.escalationPath],
    grantAll: definition.grantAll,
    permissions: [...definition.permissions],
  };
}

export function resolveAuthorizationState({ memberships = [], permissions = [] } = {}) {
  const membershipSet = new Set();
  const permissionSet = new Set();
  const permissionSources = new Map();

  const pushPermission = (permissionKey, source) => {
    const normalised = normalisePermissionKey(permissionKey);
    if (!normalised) {
      return;
    }
    permissionSet.add(normalised);
    if (!permissionSources.has(normalised)) {
      permissionSources.set(normalised, new Set());
    }
    if (source) {
      permissionSources.get(normalised).add(source);
    }
  };

  memberships
    .map(normaliseMembershipKey)
    .filter(Boolean)
    .forEach((membershipKey) => {
      const definition = membershipAliasIndex.get(membershipKey);
      if (!definition) {
        return;
      }
      membershipSet.add(definition.key);
      if (definition.grantAll) {
        permissionIndex.forEach((_, permissionKey) => pushPermission(permissionKey, definition.key));
        return;
      }
      definition.permissions.forEach((permissionKey) => pushPermission(permissionKey, definition.key));
    });

  permissions
    .map(normalisePermissionKey)
    .filter(Boolean)
    .forEach((permissionKey) => pushPermission(permissionKey, 'explicit'));

  expandPermissionSet(permissionSet, permissionSources);

  const membershipKeys = [...membershipSet];
  const membershipDetails = membershipKeys
    .map((membershipKey) => cloneMembership(canonicalMembershipIndex.get(membershipKey)))
    .filter(Boolean);

  const permissionDetails = [...permissionSet].map((permissionKey) => {
    const metadata = permissionIndex.get(permissionKey) ?? {
      key: permissionKey,
      label: permissionKey,
      description: '',
      category: 'general',
      surfaces: [],
      implies: [],
      escalationPath: [],
    };
    return {
      ...metadata,
      key: permissionKey,
      grantedBy: [...(permissionSources.get(permissionKey) ?? [])],
    };
  });

  const breakdown = {};
  permissionDetails.forEach((detail) => {
    breakdown[detail.key] = {
      metadata: detail,
      grantedBy: detail.grantedBy,
    };
  });

  return {
    version: permissionMatrix.version,
    membershipKeys,
    membershipDetails,
    permissionKeys: [...permissionSet],
    permissionDetails,
    breakdown,
  };
}

export function hasPermission(context, permissionKey, { allowAdminOverride = true } = {}) {
  const normalised = normalisePermissionKey(permissionKey);
  if (!normalised) {
    return false;
  }
  const state = context?.permissionKeys ? context : resolveAuthorizationState(context);
  const permissionSet = new Set(state.permissionKeys);
  if (permissionSet.has(normalised)) {
    return true;
  }
  if (!allowAdminOverride) {
    return false;
  }
  return state.membershipKeys.some((membershipKey) => {
    const definition = canonicalMembershipIndex.get(membershipKey);
    return Boolean(definition?.grantAll);
  });
}

export function listMembershipsForPermission(permissionKey) {
  const normalised = normalisePermissionKey(permissionKey);
  if (!normalised) {
    return [];
  }
  const allowed = [];
  membershipOrder.forEach((membershipKey) => {
    const definition = canonicalMembershipIndex.get(membershipKey);
    if (!definition) {
      return;
    }
    if (definition.grantAll || definition.permissions.has(normalised)) {
      allowed.push(cloneMembership(definition));
    }
  });
  return allowed;
}

export function getMembershipMetadata(membershipKey) {
  const normalised = normaliseMembershipKey(membershipKey);
  if (!normalised) {
    return null;
  }
  const definition = membershipAliasIndex.get(normalised);
  if (!definition) {
    return null;
  }
  return cloneMembership(definition);
}

export function describePermissionRequirement(permissionKey) {
  const metadata = permissionIndex.get(normalisePermissionKey(permissionKey));
  const allowedMemberships = listMembershipsForPermission(permissionKey);
  if (!metadata) {
    const fallback = normalisePermissionKey(permissionKey);
    return {
      permission: {
        key: fallback,
        label: fallback ?? 'requested permission',
        description: '',
        category: 'general',
        surfaces: [],
        implies: [],
        escalationPath: [],
      },
      allowedMemberships,
      message: 'You do not have permission to access this resource.',
    };
  }
  const readableRoles = allowedMemberships.map((membership) => membership.label);
  const message = readableRoles.length
    ? `Access to ${metadata.label} is limited to ${readableRoles.join(', ')}.`
    : `Access to ${metadata.label} is restricted to authorised Gigvora teams.`;
  return {
    permission: metadata,
    allowedMemberships,
    message,
  };
}

export const permissionMatrixVersion = {
  version: permissionMatrix.version,
  publishedAt: permissionMatrix.publishedAt,
  reviewCadenceDays: permissionMatrix.reviewCadenceDays,
};

export default {
  permissionMatrixVersion,
  resolveAuthorizationState,
  hasPermission,
  listMembershipsForPermission,
  describePermissionRequirement,
  normaliseMembershipKey,
  normalisePermissionKey,
  getMembershipMetadata,
};

File: gigvora-flutter-phoneapp/lib/core/permission_matrix.dart
class PermissionDefinition {
  const PermissionDefinition({
    required this.key,
    required this.label,
    this.implies = const <String>[],
    this.escalationPath = const <String>[],
  });

  final String key;
  final String label;
  final List<String> implies;
  final List<String> escalationPath;
}

class MembershipDefinition {
  const MembershipDefinition({
    required this.key,
    required this.label,
    this.aliases = const <String>[],
    this.permissions = const <String>[],
    this.tier = 'standard',
    this.grantAll = false,
  });

  final String key;
  final String label;
  final List<String> aliases;
  final List<String> permissions;
  final String tier;
  final bool grantAll;
}

class PermissionRequirement {
  const PermissionRequirement({
    required this.permission,
    required this.allowedMemberships,
    required this.message,
  });

  final PermissionDefinition permission;
  final List<MembershipDefinition> allowedMemberships;
  final String message;
}

class AuthorizationState {
  const AuthorizationState({
    required this.membershipKeys,
    required this.permissionKeys,
  });

  final Set<String> membershipKeys;
  final Set<String> permissionKeys;
}

String normaliseMembershipKey(String? value) {
  if (value == null || value.trim().isEmpty) {
    return '';
  }
  return value.trim().toLowerCase().replaceAll(RegExp(r'[^a-z0-9]+'), '_');
}

String normalisePermissionKey(String? value) {
  if (value == null || value.trim().isEmpty) {
    return '';
  }
  return value.trim().toLowerCase();
}

final Map<String, PermissionDefinition> _permissionIndex = <String, PermissionDefinition>{
  'rbac:matrix:view': const PermissionDefinition(
    key: 'rbac:matrix:view',
    label: 'View RBAC matrix',
  ),
  'rbac:matrix:audit': const PermissionDefinition(
    key: 'rbac:matrix:audit',
    label: 'Audit RBAC decisions',
    implies: <String>['rbac:matrix:view'],
  ),
  'rbac:matrix:simulate': const PermissionDefinition(
    key: 'rbac:matrix:simulate',
    label: 'Simulate RBAC policies',
    implies: <String>['rbac:matrix:view', 'rbac:matrix:audit'],
  ),
  'calendar:view': const PermissionDefinition(
    key: 'calendar:view',
    label: 'View company calendar',
  ),
  'calendar:manage': const PermissionDefinition(
    key: 'calendar:manage',
    label: 'Manage company calendar',
    implies: <String>['calendar:view'],
  ),
  'projects:view': const PermissionDefinition(
    key: 'projects:view',
    label: 'View project workspaces',
  ),
  'projects:manage': const PermissionDefinition(
    key: 'projects:manage',
    label: 'Manage project workspaces',
    implies: <String>['projects:view'],
  ),
  'notifications:read': const PermissionDefinition(
    key: 'notifications:read',
    label: 'Read inbox notifications',
  ),
  'notifications:manage': const PermissionDefinition(
    key: 'notifications:manage',
    label: 'Manage notification broadcasts',
    implies: <String>['notifications:read'],
  ),
};

final List<MembershipDefinition> _membershipDefinitions = <MembershipDefinition>[
  const MembershipDefinition(
    key: 'platform_admin',
    label: 'Platform Administrator',
    aliases: <String>['admin', 'platform:admin'],
    permissions: <String>['rbac:matrix:view', 'rbac:matrix:audit', 'rbac:matrix:simulate', 'calendar:manage', 'projects:manage', 'notifications:manage'],
    tier: 'privileged',
    grantAll: true,
  ),
  const MembershipDefinition(
    key: 'security_officer',
    label: 'Security Officer',
    aliases: <String>['security', 'security_officer'],
    permissions: <String>['rbac:matrix:view', 'rbac:matrix:audit', 'calendar:view', 'projects:view'],
    tier: 'privileged',
  ),
  const MembershipDefinition(
    key: 'workspace_admin',
    label: 'Workspace Administrator',
    aliases: <String>['workspace_admin', 'company_admin', 'agency_admin'],
    permissions: <String>['calendar:manage', 'projects:manage', 'notifications:manage', 'rbac:matrix:view'],
    tier: 'elevated',
  ),
  const MembershipDefinition(
    key: 'operations_lead',
    label: 'Operations Lead',
    aliases: <String>['operations_lead', 'operations'],
    permissions: <String>['calendar:manage', 'projects:manage', 'notifications:read'],
    tier: 'elevated',
  ),
  const MembershipDefinition(
    key: 'project_manager',
    label: 'Project Manager',
    aliases: <String>['project_manager', 'project_management'],
    permissions: <String>['projects:manage', 'calendar:view', 'notifications:read'],
  ),
  const MembershipDefinition(
    key: 'company_member',
    label: 'Company Member',
    aliases: <String>['company', 'company_member', 'company_manager'],
    permissions: <String>['calendar:view', 'projects:view', 'notifications:read'],
  ),
  const MembershipDefinition(
    key: 'agency_member',
    label: 'Agency Member',
    aliases: <String>['agency'],
    permissions: <String>['calendar:view', 'projects:view', 'notifications:read'],
  ),
  const MembershipDefinition(
    key: 'mentor',
    label: 'Mentor',
    aliases: <String>['mentor'],
    permissions: <String>['calendar:view', 'notifications:read'],
    tier: 'limited',
  ),
];

final Map<String, MembershipDefinition> _canonicalMembershipIndex = <String, MembershipDefinition>{};
final Map<String, MembershipDefinition> _membershipAliasIndex = <String, MembershipDefinition>{};

void _initialiseMembershipIndexes() {
  if (_canonicalMembershipIndex.isNotEmpty) {
    return;
  }
  for (final MembershipDefinition definition in _membershipDefinitions) {
    _canonicalMembershipIndex[definition.key] = definition;
    final Set<String> aliases = <String>{definition.key, ...definition.aliases.map(normaliseMembershipKey)};
    for (final String alias in aliases) {
      if (alias.isEmpty) {
        continue;
      }
      _membershipAliasIndex[alias] = definition;
    }
  }
}

AuthorizationState resolveAuthorizationState({
  Iterable<String> memberships = const <String>[],
  Iterable<String> permissions = const <String>[],
}) {
  _initialiseMembershipIndexes();
  final Set<String> membershipKeys = <String>{};
  final Set<String> permissionKeys = <String>{};

  void addPermission(String permission, {String? source}) {
    final String normalised = normalisePermissionKey(permission);
    if (normalised.isEmpty) {
      return;
    }
    permissionKeys.add(normalised);
    final PermissionDefinition? definition = _permissionIndex[normalised];
    if (definition != null) {
      for (final String implied in definition.implies) {
        permissionKeys.add(implied);
      }
    }
  }

  for (final String membership in memberships) {
    final String key = normaliseMembershipKey(membership);
    if (key.isEmpty) {
      continue;
    }
    final MembershipDefinition? definition = _membershipAliasIndex[key];
    if (definition == null) {
      continue;
    }
    membershipKeys.add(definition.key);
    if (definition.grantAll) {
      permissionKeys.addAll(_permissionIndex.keys);
      continue;
    }
    for (final String permission in definition.permissions) {
      addPermission(permission, source: definition.key);
    }
  }

  for (final String permission in permissions) {
    addPermission(permission, source: 'explicit');
  }

  return AuthorizationState(membershipKeys: membershipKeys, permissionKeys: permissionKeys);
}

PermissionRequirement describePermissionRequirement(String permissionKey) {
  _initialiseMembershipIndexes();
  final String normalised = normalisePermissionKey(permissionKey);
  final PermissionDefinition definition =
      _permissionIndex[normalised] ?? PermissionDefinition(key: normalised, label: normalised);
  final List<MembershipDefinition> allowed = <MembershipDefinition>[];
  for (final MembershipDefinition membership in _membershipDefinitions) {
    if (membership.grantAll || membership.permissions.contains(normalised)) {
      allowed.add(membership);
    }
  }
  final String message = allowed.isEmpty
      ? 'You do not have permission to access this resource.'
      : 'Access to ${definition.label} is limited to ${allowed.map((MembershipDefinition item) => item.label).join(', ')}.';
  return PermissionRequirement(permission: definition, allowedMemberships: allowed, message: message);
}

MembershipDefinition? membershipMetadata(String membershipKey) {
  _initialiseMembershipIndexes();
  final String normalised = normaliseMembershipKey(membershipKey);
  if (normalised.isEmpty) {
    return null;
  }
  return _membershipAliasIndex[normalised];
}
